market_env/environment.py

```diff
@@ -6,23 +6,82 @@
 
 
 from __future__ import annotations
-from typing import List
 
-class User:
-    def __init__(self, user_id: str):
-        self.user_id = user_id
 
-    def create_transaction(
-        self, transaction_id: int, recipient: str, amount: float, gas_price: float,
-        gas: int, timestamp: int, mempool: 'Mempool'
+class Chain:
+    def __init__(
+        self,
+        users: dict[str, User] = None,
+        builders: dict[str, Builder] = None,
+        proposers: dict[str, Proposer] = None,
+        mempools: dict[str, Mempool] = None,
     ):
+        if users is None:
+            users = {}
+        if builders is None:
+            builders = {}
+        if proposers is None:
+            proposers = {}
+        if mempools is None:
+            mempools = {}
+
+        self.users = users
+        self.builders = builders
+        self.proposers = proposers
+        self.mempools = mempools
+        self.blocks: list[str] = []
+
+    def add_block(self, block: str):
+        self.blocks.append(block)
+        self.update_mempools(block)
+
+    def update_mempools(self, block: str):
+        # Assuming block.transactions is a list of transaction ids
+        for mempool in self.mempools.values():
+            mempool.transactions = [t for t in mempool.transactions if t.transaction_id
+                                    not in block.transactions]
+
+class Node:
+    """Nodes include proposers, builders, and users."""
+    def __init__(self, peers: list[Node] = None):
+        self.mempool = Mempool()
+        self.peers = peers if peers is not None else []
+
+    def add_peer(self, peer):
+        """Add a peer node to the node's peer list."""
+        self.peers.append(peer)
+
+    def validate_transaction(self, transaction):
+        return transaction.amount > 0 and transaction.gas_price > 0 and transaction.gas > 0
+
+    def receive_transaction(self, transaction):
+        if self.validate_transaction(transaction):
+            self.mempool.add_transaction(transaction)
+            self.broadcast_transaction(transaction)
+
+    def broadcast_transaction(self, transaction):
```

Comment:
please have type hints overall -- what is the type for `transaction` here? for other inputs as well

if the function does not return anything, please put `-> None`